[IMP] sale_coupon_multi_use: m2m relation

oerp-odoo opened the pull request:
`coupon_multi_use_ids` on sale.order is now used to track multi-use
coupons on sale orders. `applied_coupon_ids` field did not suit multi-use
coupons case, because it is o2m field where if same coupon is applied
on next sale order, its relation with old sale order is removed.

This relation follows similar logis as is for `applied_coupon_ids` field
relation logic.

oerp-odoo committed:
[IMP] sale_coupon_multi_use: m2m relation

`coupon_multi_use_ids` on sale.order is now used to track multi-use
coupons on sale orders. `applied_coupon_ids` field did not suit multi-use
coupons case, because it is o2m field where if same coupon is applied
on next sale order, its relation with old sale order is removed.

This relation follows similar logis as is for `applied_coupon_ids` field
relation logic.

jcoux commented on sale_coupon_multi_use/i18n/fr.po:
```diff
@@ -20,6 +20,11 @@ msgstr ""
+#. module: sale_coupon_multi_use
+#: model:ir.model.fields,field_description:sale_coupon_multi_use.field_sale_coupon__sale_multi_use_ids
+msgid "Applied on Orders"
+msgstr ""
```
```suggestion
msgstr "Appliqué sur les bons de commandes"
```

jcoux commented on sale_coupon_multi_use/i18n/fr.po:
```diff
@@ -115,6 +121,12 @@ msgid ""
+#: code:addons/sale_coupon_multi_use/models/sale_coupon.py:0
+#, python-format
+msgid "Multi-Use Coupon is already applied for the same reward"
+msgstr ""
```
```suggestion
msgstr "Le coupon a déjà été appliqué sur la même récompense"
```

jcoux commented on sale_coupon_multi_use/i18n/fr.po:
```diff
@@ -133,7 +145,7 @@ msgstr "Bons de réduction"
 #: model:ir.model,name:sale_coupon_multi_use.model_sale_coupon_apply_code
 msgid "Sales Coupon Apply Code"
-msgstr "Code du bon de réduction"
+msgstr "Codice di applicazione coupon di vendita"
```
```suggestion
msgstr "Code du bon de réduction"
```

jcoux commented on sale_coupon_multi_use/models/sale_coupon.py:
```diff
@@ -40,6 +49,14 @@ def _compute_discount_fixed_amount_delta(self):
 
+    def _check_coupon_code(self, order):
+        # Same check as is for `applied_coupon_ids` field.
+        if self.program_id in order.coupon_multi_use_ids.mapped("program_id"):
```
You don't need to load `coupon_multi_use_ids` to use the mapped function.
Let Odoo manage the mapped entirely.
```suggestion
        if self.program_id in order.mapped("coupon_multi_use_ids.program_id"):
```

oerp-odoo replied:
Is there a difference?

jcoux replied:
Yes, you let Odoo manage data loading.
In your case, you load all `coupon_multi_use_ids` to access to programs.
In the full mapped, you delegate this to Odoo, and maybe it's better optimized, maybe not, but it's the responsibility of Odoo, and if in a future patch they optimize or changed something in the mapped implementation, you will benefit of it.

oerp-odoo replied:
Hm, but this was copy/pasted from this `self.program_id in order.applied_coupon_ids.mapped('program_id'):` (original check) Just changed field used, but case is basically the same.

jcoux replied:
The Odoo code is not always a good reference.

jcoux commented on sale_coupon_multi_use/models/sale_order.py:
```diff
@@ -1,32 +1,134 @@
+    def _trigger_multi_use_coupons(self, state, filter_func=None):
+        self.ensure_one()
+        # NOTE. This method must be called with ORDER_CTX_KEY on order.
+        multi_use_coupons = self._get_multi_use_coupons_to_handle()
```
```suggestion
        multi_use_coupons = (
            self.coupon_multi_use_ids - self.applied_coupon_ids
        )
```

jcoux commented on sale_coupon_multi_use/models/sale_order.py:
```diff
@@ -1,32 +1,134 @@
+        readonly=True,
+    )
+
+    def _get_multi_use_coupons_to_handle(self, filter_func=None):
+        self.ensure_one()
+        coupons = self.coupon_multi_use_ids - self.applied_coupon_ids
+        if filter_func:
+            return filter_func(coupons)
+        return coupons
```
Remove this useless function used only once time and with a parameter never used.

oerp-odoo replied:
Yes, that param is leftover from previous iteration. It can be removed. Though function usefulness should not be defined how many times its called. Its other purpose is SoC. Though I guess in this case, we can remove it.

jcoux commented on sale_coupon_multi_use/models/sale_order.py:
```diff
@@ -1,32 +1,134 @@
+            return filter_func(coupons)
+        return coupons
+
+    def _trigger_multi_use_coupons(self, state, filter_func=None):
```
Remove useless `filter_func` parameter.

oerp-odoo replied:
Leftover

jcoux commented on sale_coupon_multi_use/models/sale_order.py:
```diff
@@ -1,32 +1,134 @@
+            # Mimic same behavior as for single-user coupon.
+            order = order.with_context(**{ORDER_CTX_KEY: order})
             super(SaleOrder, order).action_confirm()
+            order._trigger_multi_use_coupons("used")
```
Use a function on sale coupon to delegate the state management, by example `use_coupon()`

jcoux commented on sale_coupon_multi_use/models/sale_order.py:
```diff
@@ -1,32 +1,134 @@
-            order = order.with_context(coupon_sale_order=order)
+            order = order = order.with_context(**{ORDER_CTX_KEY: order})
             super(SaleOrder, order).action_cancel()
+            order._trigger_multi_use_coupons("new")
```
Use a function on sale coupon to delegate the state management, by example `reset_coupon_state()`

jcoux commented on sale_coupon_multi_use/models/sale_order.py:
```diff
@@ -1,32 +1,134 @@
+            coupons_multi_use = order.coupon_multi_use_ids
+            if coupons_multi_use:
+                coupons_multi_use.with_context(**{ORDER_CTX_KEY: order}).write(
+                    {"state": "used"}
```
Use a function on sale coupon to delegate the state management, by example `use_coupon()`

jcoux commented on sale_coupon_multi_use/models/sale_order.py:
```diff
@@ -1,32 +1,134 @@
+            # action_confirm and action_cancel if Odoo fixes their bug.
+            # Now single use coupons are not reset back if you put SO to
+            # draft state.
+            coupons_multi_use = order.coupon_multi_use_ids
+            if coupons_multi_use:
+                coupons_multi_use.with_context(**{ORDER_CTX_KEY: order}).write(
+                    {"state": "used"}
+                )
```
No need to test `coupons_multi_use`.
Let Odoo manage it in `write` function.
```suggestion
            order.with_context(**{ORDER_CTX_KEY: order}).coupon_multi_use_ids.write(
                {"state": "used"}
            )
```

oerp-odoo replied:
It will call write with empty recordset, which is ugly.

jcoux commented on sale_coupon_multi_use/models/sale_order.py:
```diff
@@ -1,32 +1,134 @@
+                ids.extend(cmd[2])
+            elif cmd[0] == 4:
+                ids.append(cmd[1])
+        return self.env["sale.coupon"].browse(ids)
```
A comment on this function is welcome.

jcoux commented on sale_coupon_multi_use/models/sale_order.py:
```diff
@@ -1,32 +1,134 @@
+        # Reactivate coupons related to unlinked reward line
+        for line in self.filtered(lambda line: line.is_reward_line):
+            order = line.order_id
+            _filter = _get_order_line_coupons_filter(line)
```
Remove this filter function and copy/paste the code in following filtered.
It's not easy to read like this, it will better to read/compare with your following code:
```
related_program = self.env["sale.coupon.program"].search(
    [("discount_line_product_id", "=", line.product_id.id)]
)
```
and
```
                related_program_lines |= (
                    order.order_line.filtered(
                        lambda l: l.product_id
                        == related_program.discount_line_product_id
                    )
                    - line
```

oerp-odoo replied:
Don't get it what you intend to do here. Your showed code does different thing than mine. We need to filter coupons, not related program lines.

jcoux replied:
* Remove completely the function `_get_order_line_coupons_filter` (lines 6-7).
* Rewrite line 108:
`coupons_to_detach = order.coupon_multi_use_ids.filtered(lambda r: r.program_id.discount_line_product_id == line.product_id)`
* Rewrite lines 112-114:
`coupons_to_reactivate = (coupons_to_detach - order.applied_coupon_ids.filtered(lambda r: r.program_id.discount_line_product_id == line.product_id))`

Yes, you copy/paste two times the lambda filter, but the code will be more easy to read.

oerp-odoo committed:
[REF] sale_coupon_multi_use: naming

oerp-odoo committed:
[FIX] sale_coupon_multi_use: action_cancel

Odoo now fixed the way coupons are treated once sale order is cancelled.
Now when SO is cancelled, coupons are unattached from SO. In this case
action_draft support for multi-use is no longer needed, because when
SO is cancelled, same coupon should be readded manually.

oerp-odoo committed:
[REF] sale_coupon_multi_use: copy/paste filter

jcoux commented on sale_coupon_multi_use/models/sale_order.py:
```diff
@@ -1,32 +1,116 @@
+            order.coupon_multi_use_ids = [(5,)]
         return True
+
+    def _get_coupons_from_2many_commands(self, commands):
+        """Browse coupon IDs from 2many command."""
+        ids = []
+        for cmd in commands:
+            if cmd[0] == 6:
+                ids.extend(cmd[2])
+            elif cmd[0] == 4:
+                ids.append(cmd[1])
+        return self.env["sale.coupon"].browse(ids)
+
+    def _get_multi_use_coupons_from_2many_commands(self, commands):
+        coupons = self._get_coupons_from_2many_commands(commands)
+        return coupons.filtered("multi_use")
+
+    def write(self, vals):
+        """Extend to add multi-use coupons."""
+        if vals.get("applied_coupon_ids"):
+            # We want to add multi-use coupons that were added on
+            # standard applied_coupon_ids, so such coupon is preserved (
+            # applied_coupon_ids relation is removed once same coupon is
+            # used on multiple sale orders).
+            coupons_multi_use = self._get_multi_use_coupons_from_2many_commands(
+                vals["applied_coupon_ids"]
+            )
+            if coupons_multi_use:
+                vals["coupon_multi_use_ids"] = [(6, 0, coupons_multi_use.ids)]
+        return super().write(vals)
```
For me all these 3 functions:
- _get_coupons_from_2many_commands
- _get_multi_use_coupons_from_2many_commands
- write

are too complex compared to the needs.

I think we can do more easy:
```
    def write(self, vals):
        """Extend to add multi-use coupons."""
        result = super().write(vals)
        if vals.get("applied_coupon_ids"):
            for order in self:
                order.coupon_multi_use_ids = order.applied_coupon_ids.filtered("multi_use")
        return result
```

oerp-odoo replied:
This calls write twice. Which is bad.

jcoux replied:
It's not bad if the final result is very more easy, then more easy to read/maintainable.

oerp-odoo replied:
Hm, I checked your suggestion again and actually it would work incorrectly. It would cause same problem this PR was created to solve in a first place.

Thing is you need to check what is actually done with `applied_coupon_ids`. You can't assume correct command was used.

Now because `applied_coupon_ids` is o2m field, imagine this case:

SO1: we apply multi-use coupon on applied_coupon_ids (4 or 6 cmd used). It gets applied on coupon_multi_use_ids. So far so good.
Then we apply same multi-coupon on SO2 via applied_coupon_ids. Because that field its o2m field, relation will be lost on SO1.
Now let say we also apply single-use coupon on SO1. It gets applied on applied_coupon_ids and triggers part of code you suggested. Now it finds no multi-use coupons on applied_coupon_ids, thus overwrites all multi-use coupons and causes same problem we have without this PR.

Well I guess we could then only add new ones on coupon_multi_use_ids, but I wonder if then its any easier to read (though I dont really find my part hard to read, but thats my opinion:)).

jcoux replied:
Ok I understand, then 2 things:

- Add all this explanation into the code to help people to correctly understand what you do and why you do it like this.
- `applied_coupon_ids` must always be empty when we have multi use module installed. Because the value of `applied_coupon_ids` is wrong, we don't want to keep it.

oerp-odoo replied:
applied_coupon_ids can't be kept empty, because it is also used for single use coupon (native functionality).
Though I guess what could be done, is to not apply multi-coupon IDs on `applied_coupon_ids` and instead of copying those ids, move them. So then it would simplify relation management (we would only check applied_coupon_ids once and on other parts deal with coupon_multi_use_ids)

jcoux replied:
:+1: with your proposition! :)

oerp-odoo replied:
@jcoux updated. Moving 2many commands between keys got too complicated (before write), so refactored with multiple write. Though still needed to use commands there to make sure both fields are properly updated (because we need to be aware of specific 2many commands either way).

jcoux reviewed (changes requested):
Thanks for changes.
Last comment for me.

oerp-odoo committed:
[REF] sale_coupon_multi_use: write for multi-use

jcoux reviewed (approved):
Ok for me, thanks for changes!

Please don't merge without approval of another people to have a double validation.
